File: library/src/main/java/com/ethanhua/skeleton/ViewReplacer.java
package com.ethanhua.skeleton;

import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.lang.ref.WeakReference;

/**
 * Created by ethanhua on 2017/8/2.
 */

public class ViewReplacer {
    private static final String TAG = ViewReplacer.class.getName();
    private final WeakReference<View> mSourceViewRef;
    private WeakReference<View> mTargetViewRef;
    private int mTargetViewResID = -1;
    private WeakReference<View> mCurrentViewRef;
    private ViewGroup mSourceParentView;
    private final ViewGroup.LayoutParams mSourceViewLayoutParams;
    private int mSourceViewIndexInParent = 0;
    private final int mSourceViewId;

    public ViewReplacer(View sourceView) {
        mSourceViewRef = new WeakReference<>(sourceView);
        mSourceViewLayoutParams = sourceView.getLayoutParams();
        mCurrentViewRef = new WeakReference<>(sourceView);
        mSourceViewId = sourceView.getId();
    }

    public void replace(int targetViewResID) {
        if (mTargetViewResID == targetViewResID) {
            return;
        }
        if (init()) {
            mTargetViewResID = targetViewResID;
            View sourceView = mSourceViewRef.get();
            if (sourceView != null) {
                replace(LayoutInflater.from(sourceView.getContext()).inflate(mTargetViewResID, mSourceParentView, false));
            }
        }
    }

    public void replace(View targetView) {
        View currentView = mCurrentViewRef.get();
        if (currentView == targetView) {
            return;
        }
        if (targetView.getParent() != null) {
            ((ViewGroup) targetView.getParent()).removeView(targetView);
        }
        if (init()) {
            mTargetViewRef = new WeakReference<>(targetView);
            mSourceParentView.removeView(currentView);
            targetView.setId(mSourceViewId);
            mSourceParentView.addView(targetView, mSourceViewIndexInParent, mSourceViewLayoutParams);
            mCurrentViewRef = new WeakReference<>(targetView);
        }
    }

    public void restore() {
        if (mSourceParentView != null) {
            View currentView = mCurrentViewRef.get();
            View sourceView = mSourceViewRef.get();
            if (currentView != null && sourceView != null) {
                mSourceParentView.removeView(currentView);
                mSourceParentView.addView(sourceView, mSourceViewIndexInParent, mSourceViewLayoutParams);
                mCurrentViewRef = new WeakReference<>(sourceView);
                mTargetViewRef = null;
                mTargetViewResID = -1;
            }
        }
    }

    public View getSourceView() {
        return mSourceViewRef.get();
    }

    public View getTargetView() {
        return mTargetViewRef != null ? mTargetViewRef.get() : null;
    }

    public View getCurrentView() {
        return mCurrentViewRef.get();
    }

    private boolean init() {
        if (mSourceParentView == null) {
            View sourceView = mSourceViewRef.get();
            if (sourceView != null) {
                mSourceParentView = (ViewGroup) sourceView.getParent();
                if (mSourceParentView == null) {
                    Log.e(TAG, "The source view is not attached to any view. Source View ID: " + mSourceViewId);
                    return false;
                }
                mSourceViewIndexInParent = mSourceParentView.indexOfChild(sourceView);
            }
        }
        return true;
    }
}

File: Widgets/src/main/java/com/colin/library/android/widget/skeleton/ViewReplacer.kt
package com.colin.library.android.widget.skeleton

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.colin.library.android.utils.Constants
import com.colin.library.android.utils.Log
import java.lang.ref.WeakReference

/**
 * Author:ColinLu
 * E-mail:945919945@qq.com
 * Create:2025-05-01 12:02
 *
 * Des   :ViewReplacer
 */
class ViewReplacer(sourceView: View) {
    private val mSourceViewRef: WeakReference<View?> = WeakReference<View?>(sourceView)
    private val mSourceViewId = sourceView.id
    private val mSourceViewLayoutParams: ViewGroup.LayoutParams? = sourceView.getLayoutParams()
    private var mTargetViewRef: WeakReference<View?>? = null
    private var mTargetViewResID = Constants.INVALID
    private var mCurrentViewRef: WeakReference<View?>
    private var mSourceParentView: ViewGroup? = null
    private var mSourceViewIndexInParent = Constants.INVALID

    init {
        mCurrentViewRef = WeakReference<View?>(sourceView)
    }

    fun replace(targetViewResID: Int) {
        if (mTargetViewResID == targetViewResID) {
            return
        }
        checkParent()
        if (mSourceParentView != null) {
            mTargetViewResID = targetViewResID
            val sourceView = mSourceViewRef.get() ?: return
            replace(
                LayoutInflater.from(sourceView.context)
                    .inflate(targetViewResID, mSourceParentView, false)
            )
        }
    }

    fun replace(targetView: View) {
        val currentView = mCurrentViewRef.get()
        if (currentView === targetView) {
            return
        }
        if (targetView.parent != null) {
            (targetView.parent as ViewGroup).removeView(targetView)
        }
        checkParent()
        if (mSourceParentView != null || mSourceViewIndexInParent != Constants.INVALID) {
            mTargetViewRef = WeakReference<View?>(targetView)
            mSourceParentView!!.removeView(currentView)
            targetView.setId(mSourceViewId)
            mSourceParentView!!.addView(
                targetView, mSourceViewIndexInParent, mSourceViewLayoutParams
            )
            mCurrentViewRef = WeakReference<View?>(targetView)
        }
    }

    fun restore() {
        if (mSourceParentView != null) {
            val currentView = mCurrentViewRef.get() ?: return
            val sourceView = mSourceViewRef.get() ?: return
            mSourceParentView!!.removeView(currentView)
            mSourceParentView!!.addView(
                sourceView, mSourceViewIndexInParent, mSourceViewLayoutParams
            )
            mCurrentViewRef = WeakReference<View?>(sourceView)
            mTargetViewRef = null
            mTargetViewResID = Constants.INVALID

        }
    }

    fun getSourceView(): View? {
        return mSourceViewRef.get()
    }

    fun getTargetView(): View? {
        return if (mTargetViewRef != null) mTargetViewRef!!.get() else null
    }

    fun getCurrentView(): View? {
        return mCurrentViewRef.get()
    }

    private fun checkParent() {
        if (mSourceParentView == null || mSourceViewIndexInParent == Constants.INVALID) {
            val sourceView = mSourceViewRef.get() ?: return
            mSourceParentView = sourceView.parent as? ViewGroup
            if (mSourceParentView == null) {
                Log.e("The source view is not attached to any view. Source View ID: $mSourceViewId")
                return
            }
            mSourceViewIndexInParent = mSourceParentView!!.indexOfChild(sourceView)
        }
    }
}